paderwasn/rir_simulation/rir_simulation.py

```diff
@@ -0,0 +1,525 @@
+    src_colatitude += 90
+
+    # create reference directivity pattern
+    ref_directivity_src = CardioidFamily(
+        orientation=DirectionVector(
+            azimuth=0,
+            colatitude=90,
+            degrees=True
+        ),
+        pattern_enum=pattern_dispatcher[src_directivity]
+    )
+    # get attenuation w.r.t. reference characteristic at relative offset angles
+    # of the source-microphone constellation
+    directivity_attenuation = (
+        ref_directivity_src.get_response(
+            [src_azimuth, ], [src_colatitude, ]).item())
+    return directivity_attenuation
+
+
+def _add_stochastic_rir(
+        rir_ism, rir_len, kappa, alpha, beta, distance, room_dimensions,
+        t60, directivity_attenuation, win_len_direct, sample_rate,
+        sound_speed, change_alpha, lower_limit_alpha
+):
+    """
+    Create room impulse response consisting of early reflections
+    simulated with image source method and stochastic reverberation.
+    A target DRR is calculated and the individual parts consisting of ISM part,
+    weighting function and exponentially decaying stochastic part are combined
+    and appropriately scaled to match the target DRR.
+
+    Args:
+        rir_ism (array-like):
+            Room impulse response (RIR) consisting of early reflections
+
+        alpha (float):
+            Directivity factor of acoustic source
+
+        distance (float):
+            Distance between microphone position and acoustic source
+
+        directivity_attenuation (float):
+            Source’s directional response based on source directivity,
+            azimuth and elevation of the source
+
+        change_alpha (bool):
+            If true, maximum directivity factor based on early reflections
+            is calculated to guarantee solvability
+
+        lower_limit_alpha (float):
+            Lower limit of the range where alpha factors are drawn from
+
+        All other arguments are described in "calc_rirs".
+
+    Returns:
+        (rir, alpha):
+        rir: None, if no solution could be found for given parameters
+        otherwise combined and scaled RIR
+        alpha: the directivity factor of acoustic source used for generation
+    """
+    # maximum peak of pra is shifted by 40 samples compared to physical
+    # time of arrival (TOA)
+    pra_toa_offset = 40
+
+    # calculate stochastic reverberation part of RIR
+    n = np.arange(rir_len)
+    delta = 3 * np.log(10) / t60
+    toa_direct = int(
+        np.round(distance / sound_speed * sample_rate)) + pra_toa_offset
+    raw_stochastic = np.random.normal(size=rir_len) * np.exp(
+        -delta * (n - toa_direct) / sample_rate)
+    raw_stochastic[:toa_direct + 1] = 0
+
+    lower_lim = toa_direct - win_len_direct
+    upper_lim = toa_direct + win_len_direct + 1
+
+    # adjust alpha to maintain solvability;
+    # in some cases, the RIR of image source method may be already too diffuse
```

Comment:
In general: If the comments correspond  to complete sentences, please start that sentences with capital letters.